imjoy/apps.py

```diff
@@ -0,0 +1,341 @@
+        return [
+            f"{user_id}/{app_name}"
+            for app_name in os.listdir(self.apps_dir / user_id)
+            if not app_name.startswith(".")
+        ]
+
+    async def deploy(
+        self,
+        source: str,
+        user_id: str,
+        template: Optional[str] = None,
+        app_id: Optional[str] = None,
+        overwrite: bool = False,
+    ) -> str:
+        """Deploy a server app."""
+        if template == "imjoy":
+            if not source:
+                raise Exception("Source should be provided for imjoy plugin.")
+            config = await self.plugin_parser.parsePluginCode(source)
+            if app_id and app_id != config.name:
+                raise Exception(
+                    f"You cannot specify a different id ({app_id}) "
+                    f"for ImJoy plugin, it has to be `{config.name}`."
+                )
+            app_id = config.name
+            try:
+                temp = self.jinja_env.get_template(config.type + "-plugin.html")
+                source = temp.render(
+                    script=config.script, requirements=config.requirements
+                )
+            except Exception as err:
+                raise Exception(
+                    "Failed to compile the imjoy plugin, " f"error: {err}"
+                ) from err
+        elif template:
+            temp = self.jinja_env.get_template(template)
+            source = temp.render(script=source)
+        elif not source:
+            raise Exception("Source or template should be provided.")
+
+        app_id = app_id or shortuuid.uuid()
+        if (self.apps_dir / user_id / app_id).exists() and not overwrite:
+            raise Exception(
+                f"Another app with the same id ({app_id}) "
+                f"already exists in the user's app space {user_id}."
+            )
+
+        os.makedirs(self.apps_dir / user_id / app_id, exist_ok=True)
+
+        with open(
+            self.apps_dir / user_id / app_id / "index.html", "w", encoding="utf-8"
+        ) as fil:
+            fil.write(source)
+
+        return f"{user_id}/{app_id}"
+
+    async def undeploy(self, app_id: str) -> None:
+        """Deploy a server app."""
+        if "/" not in app_id:
+            raise Exception(
+                f"Invalid app id: {app_id}, the correct format is `user-id/app-id`"
+            )
+        if (self.apps_dir / app_id).exists():
+            shutil.rmtree(self.apps_dir / app_id, ignore_errors=True)
+        else:
+            raise Exception(f"Server app not found: {app_id}")
+
+    async def start(
+        self, app_id: str, workspace: str, token: Optional[str] = None
+    ) -> dotdict:
+        """Start a server app instance."""
+        if self.browser is None:
+            raise Exception("The app controller is not ready yet")
+        # context = await self.browser.createIncognitoBrowserContext()
+        page = await self.browser.new_page()
+        page.plugin = None
```

Comment:
I've changed the plugin attribute to a public name since we want to access it from outside. We could consider adding `imjoy` to that attribute name to make it more unique. I didn't see a way to register objects on the page in the playwright api.

Reply:
OK